server/src/main/java/org/opensearch/indices/IndicesService.java

```diff
@@ -1756,14 +1758,20 @@ private BytesReference cacheShardLevelResult(
      *
      * @opensearch.internal
      */
-    static final class IndexShardCacheEntity extends AbstractIndexShardCacheEntity {
-        private static final long BASE_RAM_BYTES_USED = RamUsageEstimator.shallowSizeOfInstance(IndexShardCacheEntity.class);
+    public final class IndexShardCacheEntity extends AbstractIndexShardCacheEntity {
         private final IndexShard indexShard;
 
-        protected IndexShardCacheEntity(IndexShard indexShard) {
+        public IndexShardCacheEntity(IndexShard indexShard) {
             this.indexShard = indexShard;
         }
 
+        public IndexShardCacheEntity(StreamInput in) throws IOException {
+            Index index = in.readOptionalWriteable(Index::new);
+            int shardId = in.readVInt();
+            IndexService indexService = indices.get(index.getUUID());
+            this.indexShard = Optional.ofNullable(indexService).map(indexService1 -> indexService1.getShard(shardId)).orElse(null);
```

Comment:
I noticed that you made this inner class non-static and made the constructors public. That's a pretty big change (especially the `static` removal).

Overall, this whole change to make `IndexShardCacheEntity` serializable confuses me, since it's just serializing the `ShardId`, which is already serializable (so you could just use `indexShard.shardId().writeTo(out)`).

Does `IndexShardCacheEntity` need to change?

Reply:
We can't move away from IndexShard as being the cacheEntity which is why all these changes are required. IndexShard is needed to collect stats at a shard level as explained above.

As we can't serialize the whole IndexShard object, we needed a way to only serialize ShardId(which uniquely defines a shard) and then extract IndexShard object from it. Let me know what u think or have better way to do this.